### PR TITLE
Graduate kube-proxy config API group to beta.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
#53645 "move KubeProxyConfiguration out of componentconfig API group" has been merged. Maybe we can take that PR as what we should do to graduate kube-proxy config API group to beta? This PR graduates the API group. Anything else should we do in this step?
I think the next step is #53782 (Remove deprecated kube-proxy flags) and graduate to GA in 1.10 or later.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: #53645 #53833

**Special notes for your reviewer**:
/cc @ncdc @mtaufen 

**Release note**:

```release-note
action required: The kubeproxy config API group has graduated from alpha to beta. Please use `kubeproxy.config.k8s.io/v1beta1`, as `componentconfig/v1alpha1` is no longer available. 
```

### CODE REVIEW

**Comment by ncdc** [comment]
@kubernetes/sig-architecture-pr-reviews @mtaufen @jbeda @smarterclayton @mikedanese @luxas  @thockin @liggitt @timstclair are we comfortable going to beta at this point? For reference, here's a bit of history:

- Pre-1.7
    - kube-proxy only supports command line flags
- 1.7
    - kube-proxy now supports both command line flags and a config file
    - The command line flags are now considered deprecated
    - The config API is `componentconfig/v1alpha1`
- 1.8
    - no changes
- master/to-be-1.9
    - The config API is now 'kubeproxy.config.k8s.io/v1alpha`

If we go to beta, it essentially means that all the field names and types for beta/GA proxy-related features should not change. I call out beta/GA features because I know that the IPVS functionality is still going to be alpha for 1.9, so any config field names/types for IPVS would have a bit more flexibility while the feature is still alpha.

Thoughts?

Also @xiangpengzhao, re release notes, I think that we ultimately need 1 release note between https://github.com/kubernetes/kubernetes/pull/53645 and this PR, stating that the `apiVersion` for `KubeProxyConfiguration` has changed from `componentconfig/v1alpha1` to `kubeproxy.config.k8s.io/<version>`, where `<version>` is either `v1alpha1` or `v1beta1` pending the outcome of this PR.

**Comment by mikedanese** [comment]
We need to consider both the types and the validation before we promote this to beta. Have relevant api reviewers had a chance to review this yet?

@kubernetes/sig-network-api-reviews

**Comment by luxas** [comment]
Ping @kubernetes/sig-network-api-reviews @kubernetes/api-approvers

**Comment by mikedanese** [comment]
Do we also need to figure out https://github.com/kubernetes/kubernetes/issues/54318?

**Comment by xiangpengzhao** [comment]
>Do we also need to figure out #54318?

@mikedanese which part of #54318 do you mean?

**Comment by ncdc** [comment]
@mikedanese I don't think so - that's really about where the structs live, but we can resolve that independently (I hope).

**Comment by ncdc** [comment]
This isn't going to make 1.9. We need to update the release notes for https://github.com/kubernetes/kubernetes/pull/53645... Or @Bradamant3 do we need to hand-edit a release notes doc?

**Comment by xiangpengzhao** [comment]
@ncdc I think it'd be better to hand-edit the release notes doc because I'm afraid the PRs in 1.9 milestone may not be scanned again to collect their release notes. So can you help to edit the release notes doc for PR #53645 ? I believe you will give it a better wording than I :)
https://docs.google.com/document/d/1veHHyDH9VoNTwP6yy9u2Q7Jj1hnBJGVPv3NPHInImgU/edit#

**Comment by ncdc** [comment]
@xiangpengzhao adding suggestions to the doc now.

**Comment by xiangpengzhao** [comment]
@ncdc thanks! I update the release note of #53645 accordingly :)

**Comment by luxas** [comment]
Let's solve https://github.com/kubernetes/kubernetes/issues/54318 with @kubernetes/sig-api-machinery-feature-requests and then proceed with this @kubernetes/sig-network-feature-requests (cc @thockin) in time for v1.10

**Comment by ncdc** [comment]
I do think my changes from this commit would be useful to bring forward before we make it beta (fixing some comments, correcting the json tag for UDP idle timeout) - https://github.com/kubernetes/kubernetes/pull/52198/commits/f074b28fe9cae52871c76e2834e70975ab473f7e

**Comment by luxas** [comment]
@ncdc can you send that as a PR please?

**Comment by xiangpengzhao** [comment]
@luxas FYI: @ncdc 's commit is already in PR #52198 which needs rebase now :)

**Comment by ncdc** [comment]
I can do it separately. I don’t see the yaml PR going anywhere without
feedback.

I’m off until after the new year but can do a separate pull then.

**Comment by xiangpengzhao** [comment]
@luxas if we want that change ASAP, I can cherry-pick Andy's commit and send a PR now. @ncdc SGTY?

**Comment by luxas** [comment]
Whatever works for me, but I think all approvers are OOO until EOY, so I don't think we're in that of a rush given we start working on this as soon as folks get back to work next year.

**Comment by xiangpengzhao** [comment]
@luxas SGTM :)

**Comment by ncdc** [comment]
API changes PR: https://github.com/kubernetes/kubernetes/pull/57754

**Comment by mtaufen** [comment]
Before you do this, please make sure you don't have a similar problem to https://github.com/kubernetes/kubernetes/issues/56171, which blocked graduating kubelet config for a while.

File: pkg/proxy/apis/kubeproxyconfig/v1beta1/types.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package v1alpha1
+package v1beta1
```
Comment (mtaufen) [inline pkg/proxy/apis/kubeproxyconfig/v1beta1/types.go]: Is there a reason `omitempty` and `// +optional` aren't used at all here?

Comment (ncdc) [reply pkg/proxy/apis/kubeproxyconfig/v1beta1/types.go]: Someone from @kubernetes/sig-network-api-reviews needs to go through and identify which fields are optional/omitempty

**Comment by ncdc** [comment]
@mtaufen 
> Before you do this, please make sure you don't have a similar problem to #56171, which blocked graduating kubelet config for a while.

If you specify a config file, kube-proxy gets 100% of its config from that file + field defaulting logic. All other specified flags are ignored.

If you do not specify a config file, the flags are honored.

Is that ok?

**Comment by liggitt** [comment]
>> Before you do this, please make sure you don't have a similar problem to #56171, which blocked graduating kubelet config for a while.
>
> If you specify a config file, kube-proxy gets 100% of its config from that file + field defaulting logic. All other specified flags are ignored.
>
> If you do not specify a config file, the flags are honored.
>
> Is that ok?

yes, that's excellent. the kubelet wasn't able to get all runtime values into config

**Comment by ncdc** [comment]
After looking 1 more time, I think there is 1 flag that is honored even if you specify the config file: `--master`. IIRC this is needed for running the proxy in a daemon set?

**Comment by liggitt** [comment]
> IIRC this is needed for running the proxy in a daemon set?

shouldn't be after https://github.com/kubernetes/kubernetes/pull/51172

**Comment by ncdc** [comment]
I assume you mean https://github.com/kubernetes/kubernetes/pull/52183/, but yeah. So the `--master` flag will just get removed when all the rest are deprecated. I don't see it as a blocker to beta graduation.

**Comment by mtaufen** [comment]
```
// hostnameOverride, if non-empty, will be used as the identity instead of the actual hostname.
HostnameOverride string `json:"hostnameOverride"`
```
Fields like this, which are likely unique to a single instance of kube-proxy (since it's unique to the node kube-proxy runs on), can be problematic if you want to configure all your instances from a shared object (e.g. the same ConfigMap) down the road.

**Comment by ncdc** [comment]
@mtaufen do you have a recommendation for per-instance configuration values?

**Comment by mtaufen** [comment]
@ncdc We haven't figured out how to handle them cleanly in componentconfig yet, so for now we've just left them as part of the set of flags that didn't make Kubelet's componentconfig API.

**Comment by mtaufen** [comment]
#57962 should go in before this too

**Comment by thockin** [comment]
@mtaufen Can I lean on you on this one too?

**Comment by mtaufen** [comment]
@thockin wish I saw that ping sooner... I will try to get this asap.

**Comment by mtaufen** [comment]
Is #54318 still a prereq for this?

**Comment by mtaufen** [comment]
I'd also like to see `omitempty` and `// +optional` set correctly before this PR merges.

**Comment by xiangpengzhao** [comment]
>Someone from @kubernetes/sig-network-api-reviews needs to go through and identify which fields are optional/omitempty

>I'd also like to see omitempty and // +optional set correctly before this PR merges.

Can someone from @kubernetes/sig-network-misc help confirm this? Thanks!

**Comment by xiangpengzhao** [comment]
>Is #54318 still a prereq for this?

As per @ncdc thoughts, it's not a prereq I think.
https://github.com/kubernetes/kubernetes/pull/55234#issuecomment-351110537
https://github.com/kubernetes/kubernetes/issues/54318#issuecomment-353609154
https://github.com/kubernetes/kubernetes/issues/54318#issuecomment-360505193

**Comment by ncdc** [comment]
Correct, I think #54318 is independent and can be resolved after this goes in.

However, before this goes beta, we should decide how to handle per-instance configuration. See https://github.com/kubernetes/kubernetes/pull/55234#issuecomment-364178555 and the discussion starting at https://github.com/kubernetes/kubernetes/pull/60372#issuecomment-368644621.

**Comment by mtaufen** [comment]
wrt #54318, keep in mind that if the ClientConnectionConfiguration struct is copied to a new API group after kube-proxy's group goes beta, you'll have to maintain it in two places and rev a new version of kube-proxy's API to consume the version that has an independent group, and also have to make sure that conversions to your internal type work.

**Comment by liggitt** [comment]
> keep in mind that if the ClientConnectionConfiguration struct is copied to a new API group after kube-proxy's group goes beta, you'll have to maintain it in two places and rev a new version of kube-proxy's API to consume the version that has an independent group, and also have to make sure that conversions to your internal type work.

if it's schema compatible (and I'd expect it to be) you'd just change to reference the type in the new package (both internally and externally). this is similar to how v1.PodTemplateSpec is referenced as a field in objects in lots of different API groups (DeploymentSpec, etc)

**Comment by mtaufen** [comment]
>it's schema compatible

If this sort of change has been done before and worked, then great. Didn't want to make assumptions.

If we added proto serializations for config (not saying we will do that any time soon, but we should keep the option open), would that make such a migration more difficult (e.g. does anything change in the proto serialization if a substructure is moved to a new API group)?

**Comment by liggitt** [comment]
> If we started using proto serializations for config (not saying we will do that any time soon, but we should keep the option open), would that make such a migration more difficult (e.g. does anything change in the proto serialization if a substructure is moved to a new API group)?

I don't think so, but would need double checking. The wire format of proto is basically a recursive memory dump with numbered fields. As long as the numbers agree and encode/decode of the individual fields agrees, I expect it would be compatible.

**Comment by ncdc** [comment]
> if it's schema compatible (and I'd expect it to be) you'd just change to reference the type in the new package (both internally and externally). this is similar to how v1.PodTemplateSpec is referenced as a field in objects in lots of different API groups (DeploymentSpec, etc)

Thanks @liggitt, you beat me to replying 😄 This is what I am expecting we'll do.

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/55234)

**Comment by xiangpengzhao** [comment]
/close
This PR doesn't apply any more.

**Pull request closed without merging**